internal/ethapi/api.go

```diff
@@ -631,6 +631,11 @@ func (s *BlockChainAPI) BlockNumber() hexutil.Uint64 {
 // block numbers are also allowed.
 func (s *BlockChainAPI) GetBalance(ctx context.Context, address common.Address, blockNrOrHash rpc.BlockNumberOrHash) (*hexutil.Big, error) {
 	state, _, err := s.b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
+	if errors.Is(err, ethereum.NotFound) && s.b.HistoricalRPCService() != nil {
```

Comment:
There is a unit test failing here. I think it's an interaction around reusing ethereum.NotFound and future blocks.

Reply:
@trianglesphere I added a mock method which address that test failure here: 9e0ac5cc3

It could be argued that we shouldn't fallback to the historical RPC if the block number is _in the future_, but: 
1. that would require a larger diff to handle properly
2. there still would be no better way of handling not found hashes.